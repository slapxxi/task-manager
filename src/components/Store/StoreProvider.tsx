import { toEpochTime } from '@lib';
import fetchDatabase from '@local/services/fetchDatabase';
import saveDatabase from '@local/services/saveDatabase';
import {
  Database,
  EpochTime,
  ID,
  Project,
  StoreEntry,
  StoreProject,
  StoreTag,
  StoreTask,
  Tag,
  Task,
} from '@local/types';
import { cloneDeep } from 'lodash-es';
import find from 'lodash-es/find';
import flatten from 'lodash-es/flatten';
import includes from 'lodash-es/includes';
import map from 'lodash-es/map';
import reduce from 'lodash-es/reduce';
import uniq from 'lodash-es/uniq';
import React, { useEffect, useReducer } from 'react';
import { Provider } from './context';

enum ActionType {
  receiveData = 'RECEIVE_DATA',
  requestData = 'REQUEST_DATA',
  updateTask = 'UPDATE_TASK',
  deleteTask = 'DELETE_TASK',
  updateProject = 'UPDATE_PROJECT',
  deleteProject = 'DELETE_PROJECT',
}

type StoreAction =
  | { type: ActionType.requestData }
  | { type: ActionType.receiveData; payload: Database }
  | { type: ActionType.updateTask; payload: Task }
  | { type: ActionType.deleteTask; payload: Task }
  | { type: ActionType.updateProject; payload: Project }
  | { type: ActionType.deleteProject; payload: Project };

interface State {
  readonly tasks: StoreEntry<StoreTask>;
  readonly tags: StoreEntry<StoreTag>;
  readonly projects: StoreEntry<StoreProject>;
  readonly isLoading: boolean;
  readonly lastUpdated: EpochTime;
}

interface Props {
  initialValue?: State;
  children: React.ReactNode;
}

const defaultState: State = {
  tags: {},
  tasks: {},
  projects: {},
  isLoading: false,
  lastUpdated: 0,
};

function StoreProvider({ initialValue = defaultState, children }: Props) {
  const [state, dispatch] = useReducer<State, StoreAction>(
    storeReducer,
    cloneDeep(initialValue),
  );

  useEffect(() => {
    dispatch({ type: ActionType.requestData });
    fetchDatabase((db) => dispatch({ type: ActionType.receiveData, payload: db }));
  }, []);

  useEffect(
    () => {
      if (state.lastUpdated === 0) {
        return;
      }
      saveDatabase({ tasks: state.tasks, tags: state.tags, projects: state.projects });
    },
    [state.lastUpdated],
  );

  return <Provider value={{ ...state, dispatch }}>{children}</Provider>;
}

function storeReducer(state: State, action: StoreAction): State {
  switch (action.type) {
    case ActionType.requestData:
      return { ...state, isLoading: true };
    case ActionType.receiveData:
      return { ...action.payload, isLoading: false, lastUpdated: state.lastUpdated };
    case ActionType.updateTask:
      return updateTask(state, action.payload);
    case ActionType.deleteTask:
      return deleteTask(state, action.payload);
    case ActionType.updateProject:
      return updateProject(state, action.payload);
    case ActionType.deleteProject:
      return deleteProject(state, action.payload);
    default:
      return state;
  }
}

function updateTask(state: State, task: Task): State {
  const { tags: updatedTags = {} } = task.tags.reduce(
    (updatedState, tag) => updateTag(updatedState, tag),
    state,
  );
  return {
    ...state,
    tags: updatedTags,
    tasks: {
      ...state.tasks,
      [task.id]: normalizeTask({
        ...task,
        tags: task.tags.map((t) => findMatchingTag(state, t)),
      }),
    },
    lastUpdated: Date.now(),
  };
}

function updateTag(state: State, tag: Tag): State {
  const updatedTag = findMatchingTag(state, tag);
  return {
    ...state,
    tags: { ...state.tags, [updatedTag.id]: { name: updatedTag.name } },
  };
}

function findMatchingTag(state: State, tag: Tag): Tag {
  const matchingTag = find(
    map(state.tags, (t, id) => ({ id, name: t.name })),
    (t) =>
      t.name.toLowerCase().trim() === tag.name.toLowerCase().trim() || t.id === tag.id,
  );
  if (matchingTag) {
    return matchingTag;
  } else {
    return tag;
  }
}

function normalizeTask(task: Task): StoreTask {
  const dbTask = {
    ...task,
    description: task.description || null,
    project: task.project || null,
    tags: (task.tags as Tag[]).map((t) => t.id),
    deadline: task.deadline ? toEpochTime(task.deadline) : null,
    createdAt: toEpochTime(task.createdAt),
  };
  delete dbTask.id;
  return dbTask;
}

function updateProject(state: State, project: Project): State {
  return {
    ...state,
    projects: { ...state.projects, [project.id]: normalizeProject(project) },
    lastUpdated: Date.now(),
  };
}

function normalizeProject(project: Project): StoreProject {
  return { name: project.name };
}

function deleteTask(state: State, task: Task): State {
  const updatedState = { ...state };
  delete updatedState.tasks[task.id];
  return deleteExtranousTags(updatedState);
}

function deleteExtranousTags(state: State): State {
  const updatedTags = reduce(
    state.tags,
    (acc, tag, tagId) => {
      if (isUsedAnywhere(state, tagId)) {
        return { ...acc, [tagId]: tag };
      }
      return acc;
    },
    {},
  );
  return { ...state, tags: updatedTags };
}

function isUsedAnywhere(state: State, tagId: ID): boolean {
  const tagsUsedElsewhere = getTagsUsedInTasks(state);
  if (includes(tagsUsedElsewhere, tagId)) {
    return true;
  }
  return false;
}

function getTagsUsedInTasks(state: State): ID[] {
  return uniq(flatten(map(state.tasks, (task) => task.tags)));
}

function deleteProject(state: State, project: Project): State {
  const updatedState = { ...state, tasks: state.tasks };
  delete updatedState.projects[project.id];
  return removeProjectFromTasks(state, project);
}

function removeProjectFromTasks(state: State, project: Project): State {
  const updatedTasks = reduce(
    state.tasks,
    (acc, task, taskId) => {
      if (task.project === project.id) {
        return { ...acc, [taskId]: { ...task, project: '' } };
      }
      return { ...acc, [taskId]: task };
    },
    {},
  );
  return { ...state, tasks: updatedTasks };
}

export { ActionType, State, StoreAction };

export default StoreProvider;
