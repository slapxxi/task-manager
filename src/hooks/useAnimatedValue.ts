import { tween } from 'popmotion';
import { useEffect, useState } from 'react';

function useAnimatedValue<T>(initialValue: T): [T, (value: T) => void] {
  const [storedValue, setStoredValue] = useState(initialValue);
  const [animatedValue, setAnimatedValue] = useState(storedValue);
  useEffect(
    () => {
      const animation = tween({
        from: animatedValue as any,
        to: storedValue as any,
      });
      const activeAnimation = animation.start((v: T) => setAnimatedValue(v));
      return () => activeAnimation.stop!();
    },
    [storedValue],
  );
  return [animatedValue, setStoredValue];
}

export default useAnimatedValue;
