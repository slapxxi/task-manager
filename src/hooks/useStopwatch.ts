import { Action } from '@local/types';
import { useEffect, useReducer, useRef } from 'react';

interface State {
  isRunning: boolean;
  lapse: number;
}

enum Actions {
  start,
  pause,
  stop,
  tick,
}

function useStopwatch(): [State, () => void, () => void] {
  const [state, dispatch] = useReducer(stopwatchReducer, { isRunning: false, lapse: 0 });
  const intervalRef = useRef<NodeJS.Timeout>();

  useEffect(
    () => {
      if (state.isRunning) {
        const startTime = Date.now() - state.lapse;
        intervalRef.current = setInterval(() => {
          dispatch({ type: Actions.tick, payload: Date.now() - startTime });
        }, 0);
      }
      return () => clearInterval(intervalRef.current);
    },
    [state.isRunning],
  );

  function handleRunClick() {
    if (state.isRunning) {
      dispatch({ type: Actions.pause });
    } else {
      dispatch({ type: Actions.start });
    }
  }

  function handleClear() {
    dispatch({ type: Actions.stop });
  }

  return [{ ...state }, handleRunClick, handleClear];
}

function stopwatchReducer(state: State, action: Action<Actions>) {
  switch (action.type) {
    case Actions.start:
      return { ...state, isRunning: true };
    case Actions.pause:
      return { ...state, isRunning: false };
    case Actions.tick:
      return { ...state, lapse: state.isRunning ? action.payload : state.lapse };
    case Actions.stop:
      return { lapse: 0, isRunning: false };
    default:
      return state;
  }
}

export default useStopwatch;
