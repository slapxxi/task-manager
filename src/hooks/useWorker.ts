import { useEffect } from 'react';
import Worker from '../main.worker.ts';

const worker = new Worker();

function useWorker<Action>(onMessage: (action: Action) => void) {
  useEffect(() => {
    worker.addEventListener('message', ({ data }) => onMessage(data));
    return () => worker.removeEventListener('message', ({ data }) => onMessage(data));
  }, []);

  return worker;
}

export default useWorker;
