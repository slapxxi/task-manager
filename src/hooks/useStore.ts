import { createTask, fromEpochTime } from '@lib';
import StoreContext from '@local/components/Store/context';
import { ActionType } from '@local/components/Store/StoreProvider';
import {
  Project,
  StoreEntry,
  StoreProject,
  StoreTag,
  StoreTask,
  Tag,
  Task,
} from '@local/types';
import compact from 'lodash-es/compact';
import map from 'lodash-es/map';
import sortBy from 'lodash-es/sortBy';
import { useContext } from 'react';

function useStore() {
  const { dispatch, ...state } = useContext(StoreContext);

  const actions = {
    updateTask(task: Task) {
      dispatch({ type: ActionType.updateTask, payload: task });
    },
    deleteTask(task: Task) {
      dispatch({ type: ActionType.deleteTask, payload: task });
    },
    updateProject(project: Project) {
      dispatch({ type: ActionType.updateProject, payload: project });
    },
    deleteProject(project: Project) {
      dispatch({ type: ActionType.deleteProject, payload: project });
    },
  };

  const selectors = {
    getTags() {
      return normalizeTags(state.tags);
    },
    getTasks() {
      return normalizeTasks(state.tasks, normalizeTags(state.tags));
    },
    getProjects() {
      return normalizeProjects(
        state.projects,
        normalizeTasks(state.tasks, normalizeTags(state.tags)),
      );
    },
  };

  return { ...selectors, actions, state };
}

function normalizeTasks(tasks: StoreEntry<StoreTask>, tags: Tag[]): Task[] {
  return sortBy(
    map(tasks, (task, key) =>
      createTask({
        ...task,
        id: key,
        tags: matchTags(task, tags),
        createdAt: fromEpochTime(task.createdAt),
        deadline: task.deadline && fromEpochTime(task.deadline),
      }),
    ),
    'createdAt',
  );
}

function matchTags(task: StoreTask, tags: Tag[]): Tag[] {
  if (task.tags) {
    return compact(task.tags.map((id) => tags.filter((it) => it.id === id)[0]));
  }
  return [];
}

function normalizeTags(tags: StoreEntry<StoreTag>): Tag[] {
  return map(tags, (item, key) => ({
    id: key,
    name: item.name || 'empty tag',
  }));
}

function normalizeProjects(projects: StoreEntry<StoreProject>, tasks: Task[]): Project[] {
  return map(projects, (item, key) => ({
    id: key,
    name: item.name,
    tasks: tasks.filter((t) => t.project && t.project === key),
  }));
}

export default useStore;
