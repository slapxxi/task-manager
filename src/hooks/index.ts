import useAnimatedValue from './useAnimatedValue';
import useStore from './useStore';

export { useAnimatedValue, useStore };
