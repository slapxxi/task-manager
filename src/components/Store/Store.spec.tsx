import { createProject, createTask } from '@lib';
import { useStore } from '@local/hooks';
import { size } from 'lodash-es';
import last from 'lodash-es/last';
import React, { ReactNode } from 'react';
import { fireEvent, render } from 'react-testing-library';
import StoreProvider, { State } from './StoreProvider';

jest.mock('@local/hooks/useWorker');

const mockStore = {
  tasks: {
    'test-task': {
      title: 'Test Task',
      description: 'Testing a task.',
      createdAt: 300,
      completed: false,
      project: 'test-project',
      subtasks: [],
      tags: ['css', 'html'],
      deadline: 200,
    },
    'second-task': {
      title: 'Second Task',
      description: 'Testing a task.',
      createdAt: 500,
      completed: false,
      project: 'second-project',
      subtasks: [],
      tags: ['html'],
      deadline: 300,
    },
  },
  tags: {
    css: {
      name: 'CSS',
    },
    html: {
      name: 'HTML',
    },
  },
  projects: {
    'test-project': {
      name: 'Test Project',
    },
    'second-project': {
      name: 'Second Test Project',
    },
  },
  isLoading: false,
  isSyncing: false,
  lastUpdated: 0,
};

function Subject({ children }: { children: (store: any) => ReactNode }) {
  const store = useStore();
  return <div>{children(store)}</div>;
}

it('provides tasks selector', () => {
  const spy = jest.fn(() => 'div');

  render(
    <StoreProvider initialValue={mockStore}>
      <Subject>{({ getTasks }) => spy(getTasks())}</Subject>
    </StoreProvider>,
  );

  expect(spy).toHaveBeenCalledTimes(1);
  expect(last(spy.mock.calls)).toMatchSnapshot();
});

it('provides tags selector', () => {
  const spy = jest.fn(() => 'div');

  render(
    <StoreProvider initialValue={mockStore}>
      <Subject>{({ getTags }) => spy(getTags())}</Subject>
    </StoreProvider>,
  );

  expect(spy).toHaveBeenCalledTimes(1);
  expect(last(spy.mock.calls)).toMatchSnapshot();
});

it('provides projects selector', () => {
  const spy = jest.fn(() => 'div');

  render(
    <StoreProvider initialValue={mockStore}>
      <Subject>{({ getProjects }) => spy(getProjects())}</Subject>
    </StoreProvider>,
  );

  expect(spy).toHaveBeenCalledTimes(1);
  expect(last(spy.mock.calls)).toMatchSnapshot();
});

it('provides raw state', () => {
  const spy = jest.fn(() => 'div');

  render(
    <StoreProvider initialValue={mockStore}>
      <Subject>{({ state }) => spy(state)}</Subject>
    </StoreProvider>,
  );

  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith(mockStore);
});

it('provides action to create tasks', () => {
  Date.now = () => 100;
  const spy = jest.fn(() => 'div');
  const Tree = (
    <StoreProvider initialValue={mockStore}>
      <Subject>
        {({ actions, state }) =>
          spy(state) && (
            <button
              onClick={() =>
                actions.updateTask(
                  createTask({
                    id: 'new-task',
                    title: 'new',
                    tags: [{ id: 'new', name: 'ruby' }, { id: 'old', name: 'css' }],
                  }),
                )
              }
              data-testid="trigger"
            />
          )
        }
      </Subject>
    </StoreProvider>
  );
  const { getByTestId, rerender } = render(Tree);

  fireEvent.click(getByTestId('trigger'));

  const latestState = getLatestState(spy);
  const newTask = latestState.tasks['new-task'];
  expect(spy).toHaveBeenCalledTimes(2);
  expect(newTask).not.toBeUndefined();
  expect(newTask.tags).toEqual(['new', 'css']);
  expect(newTask.completed).toEqual(false);
  expect(latestState.tags).toEqual({
    css: { name: 'CSS' },
    html: { name: 'HTML' },
    new: { name: 'ruby' },
  });

  rerender(Tree);
});

it('provides action to update tasks', () => {
  const spy = jest.fn(() => 'div');

  const { getByTestId } = render(
    <StoreProvider initialValue={mockStore}>
      <Subject>
        {({ actions, getTasks, state }) =>
          spy(state) && (
            <button
              onClick={() =>
                actions.updateTask({ ...getTasks()[0], title: 'Changed Task', tags: [] })
              }
              data-testid="trigger"
            />
          )
        }
      </Subject>
    </StoreProvider>,
  );

  fireEvent.click(getByTestId('trigger'));

  const latestState = getLatestState(spy);
  expect(spy).toHaveBeenCalledTimes(2);
  expect(latestState.tasks['test-task'].title).toEqual('Changed Task');
  expect(latestState.tasks['test-task'].tags).toEqual([]);
  expect(latestState.tags).toEqual({ html: { name: 'HTML' } });
});

it('provides action to delete tasks', () => {
  const spy = jest.fn(() => 'div');

  const { getByTestId } = render(
    <StoreProvider initialValue={mockStore}>
      <Subject>
        {({ actions, getTasks, state }) =>
          spy(state) && (
            <button
              onClick={() => actions.deleteTask(getTasks()[0])}
              data-testid="trigger"
            />
          )
        }
      </Subject>
    </StoreProvider>,
  );

  fireEvent.click(getByTestId('trigger'));

  const latestState = getLatestState(spy);
  expect(spy).toHaveBeenCalledTimes(2);
  expect(size(latestState.tasks)).toEqual(1);
  expect(latestState.tasks['test-task']).toBeUndefined();
  expect(latestState.tags).toEqual({ html: { name: 'HTML' } });
});

it('provides action to create projects', () => {
  const spy = jest.fn(() => 'div');

  const { getByTestId } = render(
    <StoreProvider initialValue={mockStore}>
      <Subject>
        {({ actions, state }) =>
          spy(state) && (
            <button
              onClick={() =>
                actions.updateProject(
                  createProject({ id: 'new-project', name: 'New Project' }),
                )
              }
              data-testid="trigger"
            />
          )
        }
      </Subject>
    </StoreProvider>,
  );

  fireEvent.click(getByTestId('trigger'));

  const latestState = getLatestState(spy);
  expect(spy).toHaveBeenCalledTimes(2);
  expect(size(latestState.projects)).toEqual(3);
  expect(latestState.projects['new-project']).not.toBeUndefined();
  expect(latestState.projects['new-project'].name).toEqual('New Project');
});

it('provides action to delete projects', () => {
  const spy = jest.fn(() => 'div');

  const { getByTestId } = render(
    <StoreProvider initialValue={mockStore}>
      <Subject>
        {({ actions, state, getProjects }) =>
          spy(state) && (
            <button
              onClick={() => actions.deleteProject(getProjects()[0])}
              data-testid="trigger"
            />
          )
        }
      </Subject>
    </StoreProvider>,
  );

  fireEvent.click(getByTestId('trigger'));

  const latestState = getLatestState(spy);
  expect(spy).toHaveBeenCalledTimes(2);
  expect(size(latestState.projects)).toEqual(1);
  expect(latestState.tasks['test-task'].project).toEqual('');
  expect(latestState.tasks['second-task'].project).toEqual('second-project');
});

function getLatestState(mock: jest.Mock) {
  return (last(mock.mock.calls)![0] as unknown) as State;
}
