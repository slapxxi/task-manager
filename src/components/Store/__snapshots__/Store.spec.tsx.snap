// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provides projects selector 1`] = `
Array [
  Array [
    Object {
      "id": "test-project",
      "name": "Test Project",
      "tasks": Array [
        Object {
          "completed": false,
          "createdAt": 2010-01-01T00:00:00.300Z,
          "deadline": 2010-01-01T00:00:00.300Z,
          "description": "Testing a task.",
          "id": "test-task",
          "project": "test-project",
          "subtasks": Array [],
          "tags": Array [
            Object {
              "id": "css",
              "name": "CSS",
            },
            Object {
              "id": "html",
              "name": "HTML",
            },
          ],
          "title": "Test Task",
        },
      ],
    },
    Object {
      "id": "second-project",
      "name": "Second Test Project",
      "tasks": Array [
        Object {
          "completed": false,
          "createdAt": 2010-01-01T00:00:00.300Z,
          "deadline": 2010-01-01T00:00:00.300Z,
          "description": "Testing a task.",
          "id": "second-task",
          "project": "second-project",
          "subtasks": Array [],
          "tags": Array [
            Object {
              "id": "html",
              "name": "HTML",
            },
          ],
          "title": "Second Task",
        },
      ],
    },
  ],
]
`;

exports[`provides tags selector 1`] = `
Array [
  Array [
    Object {
      "id": "css",
      "name": "CSS",
    },
    Object {
      "id": "html",
      "name": "HTML",
    },
  ],
]
`;

exports[`provides tasks selector 1`] = `
Array [
  Array [
    Object {
      "completed": false,
      "createdAt": 2010-01-01T00:00:00.300Z,
      "deadline": 2010-01-01T00:00:00.300Z,
      "description": "Testing a task.",
      "id": "test-task",
      "project": "test-project",
      "subtasks": Array [],
      "tags": Array [
        Object {
          "id": "css",
          "name": "CSS",
        },
        Object {
          "id": "html",
          "name": "HTML",
        },
      ],
      "title": "Test Task",
    },
    Object {
      "completed": false,
      "createdAt": 2010-01-01T00:00:00.300Z,
      "deadline": 2010-01-01T00:00:00.300Z,
      "description": "Testing a task.",
      "id": "second-task",
      "project": "second-project",
      "subtasks": Array [],
      "tags": Array [
        Object {
          "id": "html",
          "name": "HTML",
        },
      ],
      "title": "Second Task",
    },
  ],
]
`;
